cpp/src/link_test.cpp

```diff
@@ -23,9 +23,121 @@
 #include <proton/sender_options.hpp>
 #include <proton/receiver_options.hpp>
 #include <proton/container.hpp>
+#include <proton/connection.hpp>
+#include <proton/connection_options.hpp>
+#include <proton/listen_handler.hpp>
+#include <proton/listener.hpp>
+#include <proton/messaging_handler.hpp>
+#include <proton/types.hpp>
+#include <proton/message.hpp>
+#include <proton/target_options.hpp>
+#include <proton/source_options.hpp>
+#include <proton/delivery.hpp>
 
 #include <iostream>
 
+#include <map>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace {
+std::mutex m;
+std::condition_variable cv;
+bool listener_ready = false;
+int listener_port;
+const std::string DYNAMIC_ADDRESS = "test_dynamic_address";
+} // namespace
+
+class test_recv : public proton::messaging_handler {
+  private:
+    class listener_ready_handler : public proton::listen_handler {
+        void on_open(proton::listener &l) override {
+            {
+                std::lock_guard<std::mutex> lk(m);
+                listener_port = l.port();
+                listener_ready = true;
+            }
+            cv.notify_one();
+        }
+    };
+
+    std::string url;
+    proton::listener listener;
+    listener_ready_handler listen_handler;
+
+  public:
+    test_recv(const std::string &s) : url(s) {}
+
+    void on_container_start(proton::container &c) override {
+        listener = c.listen(url, listen_handler);
+    }
+
+    void on_receiver_open(proton::receiver& r) override {
+        std::map<proton::symbol,proton::value> m_sender({{proton::symbol("supported-dist-modes"), proton::value("move")}});
+        std::map<proton::symbol,proton::value> props = r.target().dynamic_properties();
+
+        ASSERT(r.target().dynamic());
+        ASSERT(r.target().address().empty());
+        ASSERT_EQUAL(m_sender, props);
+
+        proton::target_options opts;
+        std::map<proton::symbol, proton::value> m({{proton::symbol("supported-dist-modes"), proton::value("copy")}});
+        opts.address(DYNAMIC_ADDRESS);
+        opts.dynamic_properties(m);
+        if (r.uninitialized()) {
+            r.open(proton::receiver_options().target(opts));
+        }
+        listener.stop();
+    }
+};
+
+class test_send : public proton::messaging_handler {
+  private:
+    std::string url;
+
+  public:
+    test_send(const std::string& s) : url(s) {}
+
+    void on_container_start(proton::container& c) override {
+        proton::target_options opts;
+        std::map<proton::symbol,proton::value> m({{proton::symbol("supported-dist-modes"), proton::value("move")}});
```

Comment:
As above - maybe it would be better to have constants for move and copy symbols that can be used in this code to avoid repeating ```proton::value(proton::symbol("move"))``` etc. actually perhaps ```proton::symbol("move")``` may be enough as I think it should automatically convert to ```proton::value``

Reply:
I tried to make them 'constant' locally but I found the current version to be more readable. However, if you think adding constants will add more value, I can update it.